lib/metadata/Metadata.hpp

```diff
@@ -0,0 +1,186 @@
+/**
+ * @file
+ * @brief Contains Metadata class.
+ */
+
+#ifndef METADATA_HPP
+#define METADATA_HPP
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+#include <initializer_list>
+#include <iostream>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace chord {
+
+// The Metadata class is roughly similar to a Python dictionary. It
+// holds typed key/value pairs. The keys are strings, and the values
+// are either boolen, integer, float, string, or a vector of these.
+//
+// Values cannot be overridden once set. This is a write-once data
+// structure.
+//
+// Internally, integers and floats are stored using 64 bits.
+class Metadata {
```

Comment:
I like this as a possible replacement for the current metadata structures.   Although we will need some serialization and deserialization functions.  Doesn't necessarily need to be added now, but would be nice to have. since then we could start working on updating the network and basic file logic away form struct serialization.